File: python/graph/Solution841.py
# 841. 钥匙和房间
from typing import List


class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        # adjacency list
        graph = {}
        for i, adj in enumerate(rooms):
            graph[i] = adj
        visited = [False] * len(rooms)
        # dfs
        def dfs(node):
            if visited[node]:
                return
            visited[node] = True
            adj = graph[node]
            for room in adj:
                dfs(room)
            return
        dfs(0)
        all_visited = True
        for v in visited:
            if not v:
                all_visited = False
                break
        return all_visited

File: src/graph/Solution841.java
// 841. 钥匙和房间

package graph;

import java.util.*;

public class Solution841 {

    Map<Integer, List<Integer>> graph;
    boolean[] visited;

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        graph = new HashMap<>();
        // build adjacency list
        for (int i = 0; i < rooms.size(); i++) {
            graph.put(i, rooms.get(i));
        }
        visited = new boolean[rooms.size()];

        dfs(0);

        for (boolean v: visited) {
            if (!v) return false;
        }
        return true;
    }

    public void dfs(int room) {
        if (visited[room]) return;
        visited[room] = true;
        List<Integer> neighbors = graph.get(room);
        for (int nei: neighbors) {
            dfs(nei);
        }
        return;
    }

}
